ADR-180 (#181)
decentraland/adr: content/ADR-180-communication-adapters.md

---
layout: adr
adr: 180
title: Communication protocols
date: 2023-01-26
status: Draft # pick one of these
type: Standards Track
spdx-license: CC0-1.0
authors:
  - menduz
---

## Abstract

<!--
Abstract is a multi-sentence (short paragraph) technical summary. This should be a very terse and human-readable version of the document section. **Someone should be able to read only the abstract to get the gist of what this document is about in its current state.** Abstracts should be always up to date with the current state of the document.
-->

This document lists and describes the official communication protocols available for [realms (ADR-110)](/adr/ADR-110). This document should be updated regularly when the implementations change or are added to the explorers.

## Glossary

- **Communications transport** underlying technology and/or underlying transport protocol in which the Decentraland explorers implement an adapter

> - **Communications adapter** piece of code or object that connect the internal state of the explorer with an specific transport

Reviewer: ```suggestion
- **Communications adapter** piece of code or object that connects the internal state of the explorer with a specific transport
```

- **Realm** [set of configurations](/adr/ADR-110) for an [explorer](/adr/ADR-102)

## Solution Space Exploration

> Decentraland explorers connect to different communications transports via connection strings, those are URL-like identifiers containing all the information to select the technology and establish the connection to the target network.

Reviewer: ```suggestion
Decentraland explorers connect to different communications transports via connection strings, URL-like identifiers containing all the information to select the technology and establish the connection to the target network.
```

The basic structure of a connection string is defined with the following BNF

```bnf
PROTOCOL          := !':' [A-Za-z][A-Za-z0-9-]+
PARAMETERS        := any+
CONNECTION_STRING := PROTOCOL ':' PARAMETERS
```

Here are some examples:

```
ws-room:ws-room-service.decentraland.org/rooms/123-room-id
^^^^^^^                                                    protocol
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ parameter

offline:offline
^^^^^^^         protocol
        ^^^^^^^ parameter

livekit:wss://livekit.decentral.io?access_token=abc1223123
^^^^^^^                                                    protocol
        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ parameter

```

## Specification

> The realms will provide the ways to get a connection string for our session, with that connection string a communications adapter must be created, which will use a communications transport under the hood.

Reviewer: ```suggestion
The realms will provide ways to get a connection string for our session. A communications adapter must be created with that connection string, which will use a communications transport under the hood.
```

All communications adapters must implement the semantics defined in [ADR-104](/adr/ADR-104). Some adapters, like livekit, may take advantage of their underlying implementations to set up Voice chat.

> TODO: Document the voicechat semantics

### Resolution of communications transport

1. Get the connection string
1. Parse it and separate the components protocol and params `[protocol, params] = str.split(':', max: 2)`
1. Find the function that handles the `protocol` and call it with the `params`

### Protocols

Here is the list of officially supported protocols

- `offline` will create an empty adapter. This protocol is used to test without communications servers and by some single-player experiences like the onboarding tutorial. The parameters for the offline protocol are implementation dependent.
- `simulator` will create different simmulation conditions to evaluate performance and use cases. The parameters for the offline protocol are implementation dependent.

> - `ws-room` will create a WebSocket transport, using the wire protocol and semantics defined in [ADR-105](/adr/ADR-105). The parameter is the absolute URL to be connected to, including the hostname. If no protocol is defined in the url, `wss://` will be prepended.

Reviewer: ```suggestion
- `ws-room` will create a WebSocket transport, using the wire protocol and semantics defined in [ADR-105](/adr/ADR-105). The parameter is the absolute URL to be connected to, including the hostname. If no protocol is defined in the URL, `wss://` will be prepended.
```

- `livekit` will create a LiveKit transport. The parameter is the absolute URL to be connected to, including the hostname and the livekit token if needed. It is mandatory to specify the `wss://` protocol as part of the URL.

> - `signed-login` performs a signed-fetch as defined in [ADR-44](/adr/ADR-44). It is used to authenticate the user BEFORE sending the final connection string. The connection string comes as part of the responsein the `fixedAdapter` JSON field. In case of an error or access denied, a `message` can be sent to the user to provide more information about the reason of the denyed access.

Reviewer: ```suggestion
- `signed-login` performs a signed-fetch as defined in [ADR-44](/adr/ADR-44). It is used to authenticate the user BEFORE sending the final connection string. The connection string comes as part of the response in the `fixedAdapter` JSON field. In case of an error or denied access, a `message` can be sent to the user to provide more information about the reason for the denial.
```

#### `signed-login` verifications

> The signed-login exists as part of a "handshake" mechanism to enable authenticated log-ins and access control to different communication services like Livekit.

Reviewer: ```suggestion
The signed login exists as part of a "handshake" mechanism to enable authenticated log-ins and access control to different communication services like Livekit.
```
